tests/Analyze/Gen.hs

```diff
@@ -0,0 +1,464 @@
+
+genUnaryArithOp :: MonadGen m => m UnaryArithOp
+genUnaryArithOp = Gen.element [Negate, Abs] -- Sqrt, Ln, Exp, Signum
+
+genRoundingLikeOp :: MonadGen m => m RoundingLikeOp
+genRoundingLikeOp = Gen.element [Round, Ceiling, Floor]
+
+genComparisonOp :: MonadGen m => m ComparisonOp
+genComparisonOp = Gen.element [Gt, Lt, Gte, Lte, Eq, Neq]
+
+type ECore = Existential (Core Analyze.Term)
+
+-- TODO: we might want to generalize this to a union of open intervals
+
+-- 'NumSize' bounds the values that a generated expression can evaluate to. We
+-- use this to avoid generating large numbers in computationally expensive
+-- spots.
+type NumBound = Interval Double
+
+data BoundedType where
+  BoundedInt     :: NumBound -> BoundedType
+  BoundedDecimal :: NumBound -> BoundedType
+  BoundedString  :: Int      -> BoundedType
+  BoundedTime    ::             BoundedType
+  BoundedBool    ::             BoundedType
+  BoundedKeySet  ::             BoundedType
+  -- TODO: objects
+
+arithSize :: ArithOp -> NumBound -> (NumBound, NumBound)
+arithSize op size = case op of
+  Add -> (size, size)
+  Sub -> (size, size)
+  Mul -> (sqrt size, sqrt size)
+  Div -> (size, size)
+  Pow -> error "not yet implemented: we don't symbolically interpret this operator"
+  Log -> error "not yet implemented: we don't symbolically interpret this operator"
+
+unaryArithSize :: UnaryArithOp -> NumBound -> NumBound
+unaryArithSize op size = case op of
+  Negate -> - size
+  Sqrt   -> size ** 2
+  Abs    -> size -- ?
+  Ln     -> error "not yet implemented: we don't symbolically interpret this operator"
+  Exp    -> error "not yet implemented: we don't symbolically interpret this operator"
+  Signum -> error "not yet implemented: we don't symbolically interpret this operator"
+
+mkInt :: MonadGen m => Core Analyze.Term Integer -> m ETerm
+mkInt = pure . ESimple TInt . Inj
+
+mkDec :: MonadGen m => Numerical Analyze.Term Decimal -> m ETerm
+mkDec = pure . ESimple TDecimal . Inj . Numerical
+
+mkBool :: MonadGen m => Core Analyze.Term Bool -> m ETerm
+mkBool = pure . ESimple TBool . Inj
+
+-- | When we know what type we'll be receiving from an existential we can
+-- unsafely extract it.
+class Extract a where
+  extract :: HasCallStack => ETerm -> Analyze.Term a
+
+instance Extract Integer where
+  extract = \case
+    ESimple TInt x -> x
+    other -> error (show other)
+
+instance Extract Decimal where
+  extract = \case
+    ESimple TDecimal x -> x
+    other -> error (show other)
+
+instance Extract String where
+  extract = \case
+    ESimple TStr x -> x
+    other -> error (show other)
+
+instance Extract Bool where
+  extract = \case
+    ESimple TBool x -> x
+    other -> error (show other)
+
+instance Extract Time where
+  extract = \case
+    ESimple TTime x -> x
+    other -> error (show other)
+
+instance Extract KeySet where
+  extract = \case
+    ESimple TKeySet x -> x
+    other -> error (show other)
+
+-- TODO: Var, objects
```

Comment:
I suppose this is very related to `analyzeEval EObject{} _ = pure (Left "TODO: analyzeEval EObject")`. so nevermind the comment there now

Reply:
Yeah, this is currently limited to simple types only.